### PR TITLE
Add support for self-written flash loaders

### PR DESCRIPTION
**Pull request by Tiwalun**
Add a new sub command `extract`, which can be used to extract a flash algorithm directly from an ELF file. The existing functionality for handling CMSIS-Packs is moved to the `pack` sub command.

The goal of this change is to support self-written flash loaders, like https://github.com/Tiwalun/hifive-flashloader . The generated yaml file has to be adapted manually to enter the chip information, but the flash algorithm part is filled out automatically.

### CODE REVIEW

**Comment by Yatekii** [comment]
Yeah nice!

I will see through shortly :)

> The generated yaml file has to be adapted manually to enter the chip information, but the flash algorithm part is filled out automatically.

Maybe we could make it so that a yaml template with prefilled data can be given too as an argument and it then fills in the flash algo from the ELF :) this way no copy pasta will be required :)

File: README.md [hunk 1]
```diff
@@ -3,14 +3,27 @@
...
 
-    cargo run --  <CMSIS-PACK> out/
+    cargo run -- pack <CMSIS-PACK> out/
 
 This wil generate YAML files containing the target descriptions, which can be used with probe-rs.
 
+## Usage with ELF files
+
+The target-gen tool can also be used to create a target description based on an ELF file. This
+requires that the ELF file confirms to the ARM CMSIS standard for flash algorithms, which
```
Comment (Yatekii) [inline README.md:20]:
```suggestion
requires that the ELF file adhers to the ARM CMSIS standard for flash algorithms, which
```
I am pretty sure the use of confirm is wrong here, but I might missunderstand :)

File: src/parser.rs [hunk 2]
```diff
@@ -77,16 +80,18 @@ pub fn extract_flash_algo(
     let algorithm_binary = crate::algorithm_binary::AlgorithmBinary::new(&elf, &buffer)?;
     algo.instructions = base64::encode(&algorithm_binary.blob());
 
+    let code_section_offset = algorithm_binary.code_section.start;
```
Comment (Yatekii) [inline src/parser.rs:83]: Did you check if this works with the ARM algos? Or should I check this? :)

File: src/main.rs [hunk 3]
```diff
@@ -18,36 +18,101 @@ use probe_rs::config::{FlashRegion, MemoryRegion, RamRegion};
...
+        variants: vec![Chip {
+            name: "<chip name>".to_owned(),
+            memory_map: vec![
+                MemoryRegion::Flash(FlashRegion {
+                    is_boot_memory: false,
+                    range: 0..0x2000,
+                }),
+                MemoryRegion::Ram(RamRegion {
+                    is_boot_memory: true,
+                    range: 0x1_0000..0x2_0000,
```
Comment (Yatekii) [inline src/main.rs:92]: are we sure we want to have this hardcoded? :)

File: src/main.rs [hunk 4]
```diff
@@ -18,36 +18,101 @@ use probe_rs::config::{FlashRegion, MemoryRegion, RamRegion};
...
+
+    let chip_family = ChipFamily {
+        name: "<family name>".to_owned(),
+        core: "<mcu core>".to_owned(),
+        variants: vec![Chip {
+            name: "<chip name>".to_owned(),
+            memory_map: vec![
+                MemoryRegion::Flash(FlashRegion {
+                    is_boot_memory: false,
+                    range: 0..0x2000,
```
Comment (Yatekii) [inline src/main.rs:88]: Same here :)

File: src/main.rs (hunk 3 above)
Comment (Tiwalun) [reply src/main.rs:92]: This is just a place holder value for the generated value. I'm not sure this is the best way :wink:, but with these values you at least get a template for the structure of the YAML.

File: src/parser.rs (hunk 2 above)
Comment (Tiwalun) [reply src/parser.rs:83]: I didn't check yet. With my self-written flash loader it works :wink:

**Comment by Tiwalun** [comment]
I've added an update function to the `extract` command, and removed some of the duplicate structs.

**Pull request merged**